### ISSUE TITLE
Cannot connect to TLS 1.1/TLS 1.2 instances

### ISSUE THREAD

**Issue by fpro123** [WaelHamze/xrm-ci-framework#74]
Hi, I'm connecting succesfully to our non-trial instances (online), however I'm not having success connecting to trial Dynamics 365 CE instances (tried two different ones). 
I'm using 
AuthType=Office365;Username=xx@xxxx.onmicrosoft.com; Password=xxxx; Url=https://xxx.crm4.dynamics.com
The exact same connection string works well in XrmToolbox.
The user has the role of System Administrator.
I'm trying a simple "MSCRM Publish Customizations"-task.
Is there a bug somewhere?
Thanks for a great framework.

**Comment by T-REX-XP** [comment]
I have got the same issue with Ping/ExportSolution task. If i use ondemand D365, which connection string i have to use??

**Comment by T-REX-XP** [comment]
I found solution:
use PowerShell to improve SecurityProtocol:
[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;

or add  ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12; 
in your c# code before using CrmServiceClient

It is worked for me.

**Comment by WaelHamze** [comment]
@fpro123 @T-REX-XP thx for sharing this. Actually everything works fine for me using current code with Dynamics 365 CE trials as I developed most of the tasks using this.

What was the error that you were getting before you added the code above?

Also where are you running your builds? It could be something to do with how that server was setup and the default protocols configured.

Either way I am keen to get to the bottom of this so we can have a solution that works for online and on-premise.

**Comment by fpro123** [comment]
Hi @WaelHamze 
I'm not using thePowerShell Cmdlets, only the Build & Release tasks.
I guess the annoying thing is that my connection string works with XrmToolBox and not with the VSTS task.
I'm getting this in the log: 
2018-01-05T07:05:56.3875332Z ##[warning]Unable to Login to Dynamics CRM
2018-01-05T07:06:12.1850895Z ##[warning]Unable to Login to Dynamics CRM
2018-01-05T07:06:27.5975147Z ##[warning]Unable to Login to Dynamics CRM
2018-01-05T07:06:28.2939177Z ##[error]Couldn't connect to CRM instance after 3 attempts: Unable to Login to Dynamics CRM

**Comment by T-REX-XP** [comment]
Hi @WaelHamze 
i'm using the PowerShell Cmdlets (in future with jenkins) for Export and unpack customization from d365 online instance. Instance is not trial.

Client OS: Windows 10 (i try from home and work PC) 
No one PowerShell Cmdlets wasn't connected to D365 v9, but connected fine to  D365 v8.2
 
I have got the same error:
[warning]Unable to Login to Dynamics CRM
[warning]Unable to Login to Dynamics CRM
[warning]Unable to Login to Dynamics CRM
[error]Couldn't connect to CRM instance after 3 attempts: Unable to Login to Dynamics CRM

so my workaround i found after spending 1.5 days.

it is run(in Powershell) the following string before calling your cmdlets:    

[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;


Currently all works fine for me. Thanks for you job.

**Comment by WaelHamze** [comment]
@fpro123 @T-REX-XP 

Thanks for the prompt feedback. I guess some instances may still be running TLS 1.1/1.0 where others might be running TLS 1.2

I noticed the Cmdlets are currently compiled using .NET 4.6. As per the article below .NET 4.6.2 is required to automatically select the highest secure version.

I guess is machine is configured to use TLS 1.2 then things work fine. The code above forces the use of TLS 1.2 which makes things work.

So I'll get the Cmdlets updated to 4.6.2 and see whether I can get any more useful information out of these errors messages.

**Comment by bo-stig-christensen** [comment]
@WaelHamze 
Have you updated the VSTS Build Tasks with TLS 1.2? - I seem to get "Cannot login to CRM" error messages.
I am using the VSTS build tasks against v9.0 instances and getting this error message.

**Comment by devkeydet** [comment]
Looks like @WaelHamze fixed it in the source code 3 days ago:
https://github.com/WaelHamze/xrm-ci-framework/commit/6b442fbb20029bfb8843bbd0c6bd9a6f76c26706

However, it looks like the updated VSTS tasks haven't made it through the publishing process:https://marketplace.visualstudio.com/items?itemName=WaelHamze.xrm-ci-framework-build-tasks

As of submitting this comment, the tasks haven't been updated yet.  The "Last updated" timestamp is "‎12‎/‎4‎/‎2017‎ ‎4‎:‎14‎:‎55‎ ‎PM"

**Comment by bo-stig-christensen** [comment]
Also, in general no release as such has been created ;-)

**Comment by WaelHamze** [comment]
I can't replicate this issue actually. The existing code is working for me against all my instances and even the trial I setup about an hour ago on crm11. These are all V9 instances.

Not sure what is the impact of running the code below for users whose instances are still on TLS 1.1.
[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;

Are you testing this on the hosted Agent or your private agents?

I was testing on hosted agents perhaps the default is configured to use TLS 1.2.

I have updated the Cmdlet versions to .NET 4.6.2 which I can release to the marketplace anyway.

Let me know your thoughts?

**Comment by bo-stig-christensen** [comment]
@WaelHamze
I am using a private agent. Maybe MS has some kind of specific setup on their hosted agents. - This could definitely be the difference.
In general, the .NET 4.6.2 should do the trick, as that version uses the TLS protocol to version 1.2.

I think that if you release that .NET 4.6.2 version to VSTS it should fix the issues and also should not break anything existing..... except if someone is using a private agent, but does not have .NET 4.6.2 on the server of course ;-)

**Comment by devkeydet** [comment]
Yes, I was using the "Hosted VS2017" agent.  When I do the following locally on my PC:

Select-WhoAmI -ConnectionString "AuthType=Office365;Url=https://some-instance.crm.dynamics.com;Username=someuser@sometenant.onmicrosoft.com;Password=somePassword"

I consistently got the same "Unable to Login" error I am getting in the hosted agent.  When I call the following, it succeeds consistently:

[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;
Select-WhoAmI -ConnectionString "AuthType=Office365;Url=https://some-instance.crm.dynamics.com;Username=someuser@sometenant.onmicrosoft.com;Password=somePassword"

It succeeds.  I am going to try this on the hosted agent by manually installing your scripts from powershell gallery as a sanity check.

**Comment by devkeydet** [comment]
"Hosted VS2017" agent repro:

1. Create a new empty release (no build necessary)
2. Add a single PowerShell task and add the following inline script:

Install-Module -Name Xrm.Framework.CI.PowerShell.Cmdlets -Force -Scope CurrentUser
Select-WhoAmI -ConnectionString "AuthType=Office365;Url=https://some-instance.crm.dynamics.com;Username=someuser@sometenant.onmicrosoft.com;Password=somePassword"

You should see the same "Unable to Login" error we are seeing.  

3. Go back to the PowerShell task and add the following to the top of the inline script:

[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;

It should work.  So *if* 4.6.2 alone doesn't resolve, then adding the above line to PowerShell fbefore creating the connection should do the trick.  I've seen reports where 4.6.2 alone doesn't resolve this issue in Azure App Service because the web.config still specifies an incorrect targetFramework value in some places.  Example:

  <compilation debug="true" targetFramework="4.7" />
    <httpRuntime targetFramework="4.5.2" />

Perhaps something like this is going on in the hosted agent.

**Comment by WaelHamze** [comment]
Ok so I updated the Cmdlet DLL to .NET 4.6.2 but that didn't help either. Looks like PowerShell uses into own runtime and configuration.

It also turned out that hosted agent possibly depending on location and VM image, have different runtime settings.

Online supports mutiple TLS versions however TLS 1.2 only is or will be supported depending on your instance. On-premise still have the flexibility or option to remain on older TLS version.

Given the users of this framework will be using both we need the ability to support multiple protocols.

Looks like older runtimes have only SSL3 and TLS. Newer runtimes will have TLS 1.1 and TLS 1.2

I have added the below code to Cmdlets and PS Scripts which should hopefully fix the issue for all.

`            WriteVerbose(string.Format("Current Security Protocol: {0}", ServicePointManager.SecurityProtocol));

            if (!ServicePointManager.SecurityProtocol.HasFlag(SecurityProtocolType.Tls11))
            {
                ServicePointManager.SecurityProtocol = ServicePointManager.SecurityProtocol ^ SecurityProtocolType.Tls11;
            }
            if (!ServicePointManager.SecurityProtocol.HasFlag(SecurityProtocolType.Tls12))
            {
                ServicePointManager.SecurityProtocol = ServicePointManager.SecurityProtocol ^ SecurityProtocolType.Tls12;
            }

            WriteVerbose(string.Format("Modified Security Protocol: {0}", ServicePointManager.SecurityProtocol));`

The Market place has been updated 9.0.13
